### PR TITLE
carli/contour_flashing_during_animation

### LINKED ISSUE #579: incomplete contour rendering when stop animating
frontend: 9a74dbc
backend: 9ca101c

to reproduce the issue:
1) load a large image cube in spatial domain (in my case, 2800x2800)
2) generate contours with 99.9% clip (in my case, mostly noise)
3) enable animation at 5 fps or higher
4) try to stop the animation and we "may" see contours are rendered incompletely. If not, repeat 3) and 4)

Screen Shot 2019-10-28 at 14 59 10

### PR DESCRIPTION
**Pull request by crocka**
**Description**

This PR is to fix the flickering rendering of the contour drawing during animation. The problem is caused by rendering the contour drawing as soon as the contourData has arrived. The fix to check for 1) the progress is completed for the incoming level (progress === 1), and if all the levels requested by the user have been received. This PR closes #579.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~ / ~new e2e test created~
- [x] changelog updated / ~~no changelog update needed~~
- [x] ~~protobuf updated to the latest dev commit~~ / no protobuf update needed
- [x] ~~protobuf version bumped~~ / protobuf version not bumped
- [x] `BackendService` unchanged / ~~`BackendService` changed and corresponding ICD test fix added~~

### CODE REVIEW

**Comment by veggiesaurus** [comment]
aren't we still aiming to do progressive contour updates outside of animation?

**Comment by kswang1029** [comment]
> aren't we still aiming to do progressive contour updates outside of animation?

Uh we should. @crocka we will just need to wait contour data stream completion only when animation playback is on. For static image, it should keep the current way (ie progress undate/rendering)

File: src/components/ImageView/ContourView/ContourViewComponent.tsx [hunk 1]
```diff
@@ -25,14 +25,29 @@ export class ContourViewComponent extends React.Component<ContourViewComponentPr
     componentDidMount() {
         this.contourWebGLService = ContourWebGLService.Instance;
         this.gl = this.contourWebGLService.gl;
+        const contourStream = AppStore.Instance.backendService.contourStream;
         if (this.canvas) {
-            this.updateCanvas();
+            contourStream.subscribe(() => {
+                const animatorStore = AnimatorStore.Instance;
+                const receivedContourStores = Array.from(this.props.frame.contourStores.values());
+                if (receivedContourStores.every(contourImageData => contourImageData.progress === 1) && receivedContourStores.length === this.props.frame.contourConfig.levels.length && animatorStore.serverAnimationActive) {
```
Comment (YuHsuan-Hwang) [inline src/components/ImageView/ContourView/ContourViewComponent.tsx]: Can we simply use `FrameStore.contourProgress`?
Do we consider contours from matched images (`AppStore.contourFrames`)?

**Review by YuHsuan-Hwang — APPROVED** [review]
Looks good.

**Review by kswang1029 — APPROVED** [review]
looks good. no regression from e2e tests. 👍

File: src/stores/Frame/FrameStore.ts [hunk 2]
```diff
@@ -102,7 +102,7 @@ export class FrameStore {
     private static readonly ZoomInertiaDuration = 250;
     private static readonly CursorMovementDuration = 250;
 
-    private spectralFrame: AST.SpecFrame;
+    public spectralFrame: AST.SpecFrame;
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/FrameStore.ts]: @crocka Please change it back to private.
Please also update the changelog, thanks.

**Pull request merged**